<?xml version="1.0"?>
<launch>
  
  <!-- 起動時引数定義 -->
  <arg name="enable_drift_compensation" default="true" 
       description="ドリフト補正機能を有効にする"/>
  <arg name="enable_vibration_filtering" default="true" 
       description="振動ノイズ除去機能を有効にする"/>
  <arg name="debug_output" default="false" 
       description="デバッグ情報を出力する"/>
  
  <!-- パラメータ設定 -->
  <arg name="drift_compensation_gain" default="0.025" 
       description="ドリフト補正ゲイン (0.001-0.1)"/>
  <arg name="bias_estimation_rate" default="0.003" 
       description="バイアス推定レート (0.0001-0.01)"/>
  <arg name="zero_velocity_threshold" default="0.05" 
       description="停止判定閾値 [m/s] (0.01-0.1)"/>
  
  <arg name="lowpass_cutoff_frequency" default="12.0" 
       description="ローパスフィルタ遮断周波数 [Hz] (1.0-50.0)"/>
  <arg name="vibration_detection_threshold" default="0.6" 
       description="振動検出閾値 (0.1-2.0)"/>
  <arg name="median_filter_window_size" default="7" 
       description="メディアンフィルタ窓サイズ (3-21)"/>
  
  <!-- トピック名設定 -->
  <arg name="input_imu_topic" default="/sensing/imu/imu_data" 
       description="入力IMUトピック名"/>
  <arg name="input_odom_topic" default="/localization/kinematic_state" 
       description="入力オドメトリトピック名"/>
  <arg name="output_imu_topic" default="/sensing/imu/filtered_imu_data" 
       description="出力フィルタ済みIMUトピック名"/>
  
  <!-- メインノード起動 -->
  <node pkg="imu_advanced_filter" exec="imu_advanced_filter_node" 
        name="imu_advanced_filter" output="screen">
    
    <!-- 機能設定 -->
    <param name="enable_drift_compensation" value="$(var enable_drift_compensation)"/>
    <param name="enable_vibration_filtering" value="$(var enable_vibration_filtering)"/>
    <param name="debug_output" value="$(var debug_output)"/>
    
    <!-- ドリフト補正パラメータ -->
    <param name="drift_compensation_gain" value="$(var drift_compensation_gain)"/>
    <param name="bias_estimation_rate" value="$(var bias_estimation_rate)"/>
    <param name="zero_velocity_threshold" value="$(var zero_velocity_threshold)"/>
    
    <!-- 振動ノイズ除去パラメータ -->
    <param name="lowpass_cutoff_frequency" value="$(var lowpass_cutoff_frequency)"/>
    <param name="vibration_detection_threshold" value="$(var vibration_detection_threshold)"/>
    <param name="median_filter_window_size" value="$(var median_filter_window_size)"/>
    
    <!-- トピック名のリマップ -->
    <remap from="input/imu" to="$(var input_imu_topic)"/>
    <remap from="input/odometry" to="$(var input_odom_topic)"/>
    <remap from="output/filtered_imu" to="$(var output_imu_topic)"/>
  </node>

</launch>
